File: src/main/java/io/koju/autopos/catalog/service/ItemServiceImpl.java
package io.koju.autopos.catalog.service;

import com.mysema.query.types.Predicate;
import io.koju.autopos.catalog.domain.Item;
import io.koju.autopos.catalog.domain.QItem;
import io.koju.autopos.catalog.struct.filter.ItemFilter;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@Transactional
class ItemServiceImpl implements ItemService {

    private final static Logger log = LoggerFactory.getLogger(ItemServiceImpl.class);

    private final ItemRepository itemRepository;

    @Autowired
    ItemServiceImpl(ItemRepository itemRepository) {
        this.itemRepository = itemRepository;
    }

    @Override
    public Item save(Item item) {
        log.debug("Request to save Item : {}", item);
        return itemRepository.save(item);
    }

    @Override
    @Transactional(readOnly = true)
    public Page<Item> findAll(ItemFilter itemFilter, Pageable pageable) {
        log.debug("Request to get all Items");
        final Predicate filterPredicate = itemFilter.toQueryDslPredicate(QItem.item);
        return itemRepository.findAll(filterPredicate, pageable);
    }

    @Override
    @Transactional(readOnly = true)
    public Item findOne(Long id) {
        log.debug("Request to get Item : {}", id);
        return itemRepository.findOneWithEagerRelationships(id);
    }

    @Override
    public void delete(Long id) {
        log.debug("Request to delete Item : {}", id);
        itemRepository.delete(id);
    }
}

File: src/main/scala/io/koju/autopos/catalog/service/ItemServiceImpl.scala
package io.koju.autopos.catalog.service

import java.lang.Long

import io.koju.autopos.catalog.domain.Item
import org.slf4j.LoggerFactory
import org.springframework.data.domain.{Page, Pageable}
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
@Transactional
class ItemServiceImpl(private val itemRepository: ItemRepository)
  extends ItemService {

  private val log = LoggerFactory.getLogger(classOf[ItemServiceImpl])

  override def save(item: Item): Item = {
    log.debug("Request to save Item : {}", item)
    itemRepository.save(item)
  }

  @Transactional(readOnly = true)
  override def findAll(query: String, pageable: Pageable): Page[Item] = {
    log.debug("Request to get all Items")

    if (query.isEmpty) {
      itemRepository.findAll(pageable)
    } else {
      itemRepository.findByCodeIgnoreCaseContainingOrNameIgnoreCaseContaining(query, query, pageable)
    }
  }

  @Transactional(readOnly = true)
  override def findOne(id: Long): Item = {
    log.debug("Request to get Item : {}", id)
    itemRepository.findOneWithEagerRelationships(id)
  }

  override def delete(id: Long): Unit = {
    log.debug("Request to delete Item : {}", id)
    itemRepository.delete(id)
  }

  override def adjustQuantity(item: Item, number: Integer): Unit = {
    val dbItem = itemRepository.findOne(item.getId)
    dbItem.setQuantity(dbItem.getQuantity + number)
    itemRepository.save(dbItem)
  }
}
